### ISSUE TITLE
Discussion for Ubernetes scheduler flow

### ISSUE THREAD

**Issue by mfanjie** [kubernetes/kubernetes#23269]
I am working with Huawei PaaS team on Ubernetes Phase I implementation, and focus on Uberenetes scheduler. I created a sequence diagram to illustrate the process of ubernetes scheduler and want to start some discussions.

**A general principle design spec**
Currently clusterSelector (implemented as a LabelSelector) only supports a simple list of acceptable clusters. Workloads will be evenly distributed on these acceptable clusters in phase one. 

- [ ] Need to make agreement that "evenly distributed" simply means cloning Federation-RC as sub-RC, and assign target clusters, replicas will not be cared in Federation level, that is sub-RC will keep the same value for replicas of Federation RC. 
 
**Supported cases in phase I**
#1. User specifies clusterSelector in RC Spec, e.g. The following RC is created in Federation-1 with 3 running clusters a, b and c.
```
apiVersion: v1
kind: ReplicationController
metadata:
  name: nginx-controller
spec:
  replicas: 5
  ...
  clusterSelector:   
  name in (a, b) 
```
The result will be, if scheduling succeeds, the two sub-RC are created, each RC has 5 replicas, one RC is assigned to cluster a, the other RC is assigned to cluster b.

#2. User does not specify clusterSelector in RC Spec, e.g. The following RC is created in Federation-1 with 3 running clusters a, b and c.

- [ ] need to discuss should we support this case?

``` 
apiVersion: v1
kind: ReplicationController
metadata:
  name: nginx-controller
spec:
  replicas: 5
  ...
```
As a result, the RC will be cloned to 3 sub-RC and assigned to all running clusters a, b, and c in Federation-1.

**Code logics, details in the diagram**
- [ ] most of code of package plugin/cmd/kube-scheduler could be reused, minor changes include package import to ube-scheduler, and remove some unnecessary logics.

- [ ] The design of ube-scheduler will be created based on k8s scheduler code, but with some logical changes, highlighted in RED in the sequence diagram.

1. There is no need for binding pod, so Ubernetes scheduler will not check scheduled Federation RC, no code like scheduledPodPopulator
2. Only default predicate and priority plugins are supported in P1
3. No need to watch Federation Services, as there is no need for spreading sub-RC of same Federation Service
4. About scheduler plugins
a). There will be two predicates plugins, MatchClusterSelector and RCFitsResources
b).
 From above cases description, I did not see requirement to do prioritize, so only a dummy priority plugin will be created.
5. The plugin will return a list of clusters, the Federation RC will be cloned as sub-RC and each element of the cluster list will be assigned as target.
6. As #1 described, no need for cluster binding, scheduling done after step 5.

**Dependencies**

Need definition of cluster, federation RC and sub RC

- Cluster entity is already in WIP

- [ ] Federation RC and sub RC definition need to be settled ASAP.

- [ ] Need api of accessing new defined entities in pkg/client/cache/listers.go

![scheduler](https://cloud.githubusercontent.com/assets/16643108/13914519/d4ef1444-ef89-11e5-958b-acef4e60c9fe.png)

**Comment by huangyuqi** [comment]
> The plugin will return a list of clusters, the Federation will be cloned as sub-RC and each element of the cluster list will be assigned as target.

Hi, @mfanjie , the return result would be a list of sub-RC,and the cluster is just a field of sub-RC.

**Comment by mfanjie** [comment]
@huangyuqi what I meant here is the returned value of step 26 in the sequence diagram, the function definition will be similar like followings, it will return prioritized cluster list, and the next step 29 is to iterate the returned list and created sub-rc and assign cluster.

func PrioritizeClusters(
	federationRC *api.ReplicationController,
	clusterNameToInfo map[string]*schedulercache.ClusterInfo,
	priorityConfigs []algorithm.PriorityConfig,
	ClusterLister algorithm.ClusterLister,
	extenders []algorithm.SchedulerExtender,
) (schedulerapi.ClusterPriorityList, error)

**Comment by mfanjie** [comment]
@alfred-huangjian @quinton-hoole @davidopp please comment freely, thanks.

**Comment by zhouhaibing089** [comment]
Haven't read in detail, one thought here, using `ReplicaSet` instead of `ReplicationController`maybe better, this is not the key point though. :)

**Comment by quinton-hoole** [comment]
@mfanjie Thanks for this.  I still need to got through your sequence diagram in detail, but first I'd like to address a few elements of the higher level design in the mean time:

1. As far as I understand, the number of replicas created across all clusters in this proposal will be the requested number of replicas, times the number of clusters specified in the cluster selector (or all clusters, if that is not specified).  I don't think that this is correct, as it breaks the semantics of the API.  The number of replicas created should be equal to the number requested.  As a starting point, we agreed in the discussions in the working group that only one Kubernetes RC ("sub-RC"  in your proposal) would be created, not one in each cluster.  That will solve this problem.  After that we can consider spreading across more than one cluster.  So specifically, in your use case 1, create the rc in one of the specified clusters, and in use case 2, create it in any one of the clusters in the federation.

2. In phase 1 we should maintain compatibility with the Kubernetes API precisely, if at all possible (and I think that it is).  See the section titled "Federated Replication Controllers" in https://github.com/kubernetes/kubernetes/blob/master/docs/design/federated-services.md for some details. The only thing missing in that example is the clusterSelector.

I will add more once I've reviewed your sequence diagram in more detail.

**Comment by mfanjie** [comment]
@quinton-hoole thanks. What you described was my previous thought, saying replicas = 8 and cluster number is 3, the scheduler should split the 8 replicas to 3 clusters. But my mind was changed after I had call with @huangyuqi . As if we go this way, the logic will be much more complex, so I raised it here to discuss if it is possible to simply clone to federation RC, otherwise we need to consider the following situations:
1. Replicas=8 cluster=3, when replicas can not be exact decided by cluster numbers, where to put the extra 2 cluster. Maybe place the workload as 3 *3 as 9 replicas in total?
2. Replicas=3 cluster=8, similarly, how to prioritize clusters
3. Rescheduling is needed when users run scale up/down
The logic will be more clear based one the assumption of simply clone the RC. If it can not go that way, I need to change the dummy priority plugin in the diagram to handle above logics.

**Comment by quinton-hoole** [comment]
@mfanjie As I mentioned, in phase 1, just create all replicas in one cluster (and yes, you can just use cloning, as you suggest). In a later phase we will consider spreading across more than one cluster, and yes, in that case, unequal numbers of replicas will potentially be created in each cluster.

**Comment by mfanjie** [comment]
@quinton-hoole okey, so in Phase I, we only place the RC to one of user specified clusters, for the case that use defined clusterSelector. And place the RC to any of the running clusters if no clusterSelector is defined.
About entity change.
For Federation, I agree that only ClusterSelector is need to be added, but checking the code, this property would be appropriately added to PodSpec like NodeSelector.
```
type PodSpec struct {
	// ClusterSelector is a selector which must be true for the RC to fit on a cluster
	ClusterSelector map[string]string `json:"clusterSelector,omitempty"`
} 
```
ClusterSelector property is for persisting user input, and we would need another property for persisting scheduling result, like clusterName, PodSpec is the best struct to hold that property per my current understanding.

**Comment by quinton-hoole** [comment]
Thanks @mfanjie  - your approach looks good to me, except for clusterName, which won't work well when we progress to perhaps spreading across more than one cluster.  I think that we will more than likely expose the underlying replication controllers instead, but we have not figured out the details of that yet.  It is on my TODO list to extend the existing design, as per the discussion we had last week.  I would suggest following https://github.com/kubernetes/kubernetes/blob/master/docs/design/federated-services.md directly for phase 1 (i.e. only expose the fields that the Kubernetes RC already does), and extend that once the above design is finalized.

**Comment by mfanjie** [comment]
@quinton-hoole where you want to store the scheduling result. The scheduler will watch apiserver for new federation RC, put it into a queue, then pop up the queue, and do scheduling, create sub RC for each target cluster. The clusterName is where to save the scheduling result. When the workload need to be split to more than one clusters, one sub RC will be created for each target cluster. That's my current thought.

**Comment by mfanjie** [comment]
@quinton-hoole ok, I get the point, will check out and get back to you when I have clear idea.

**Comment by quinton-hoole** [comment]
@mfanjie I would suggest that you use labels or annotations to associate the various components (e.g. label the Ubernetes RC with the same key and value as the parent Ubernetes RC.

**Comment by quinton-hoole** [comment]
Sorry, fixed confusing typo in the original version of the above comment.

**Comment by alfred-huangjian** [comment]
@quinton-hoole 
After some discussion offline with @mfanjie，we prefer to make sure whether to support spreading replicas of RC into serveal clusters. In my opinion it's better to have this function in Phase I directly which is mentioned in federated-services.md. How about your ideas?

**Comment by quinton-hoole** [comment]
@alfred-huangjian That's fine. My priority at this point is velocity, so for example if implementing spreading across clusters is non-trivial (as aluded in https://github.com/kubernetes/kubernetes/issues/23269#issuecomment-199454061), I would rather delay that then delay getting the basic control plane up and running. Lets discuss it further at tomorrow's working group meeting.  @alfred-huangjian, @mfanjie will you be there (9:30 AM, U.S. Pacific Time).  My apologies, I know that the time is not very convenient in China :-)

**Comment by alfred-huangjian** [comment]
@quinton-hoole OK, We will attend the meeting including @mfanjie and @huangyuqi. Let's discuss it over the meeting. However we will try to build a working system asap .

**Comment by mfanjie** [comment]
scheduler wip #23420

**Comment by alfred-huangjian** [comment]
@quinton-hoole Sorry there is some signal issue with my meeting connection. We agree on implementing the simple approach of not dividing the federated RC into clusters at the first phase. Meanwhile we can have more discussion the details of implementing the further approach.

**Comment by quinton-hoole** [comment]
cc @madhusudancs @nikhiljindal FYI

**Comment by madhusudancs** [comment]
PodSpec is not the right place to put cluster specific information. As @quinton-hoole suggested we should consider putting it into annotations in the first version if that's easiest thing to do.

As described in the doc here [federation-phase-1.md#replication-controller](https://github.com/kubernetes/kubernetes/blob/master/docs/design/federation-phase-1.md#replication-controller), the cluster specific information should be in RC Spec. RC definition will look like this:

```
type ReplicationController struct {  // We should probably call this ReplicaSet now, but just using RC to reduce confusion.
  unversioned.TypeMeta

  ObjectMeta

  Spec ReplicationControllerSpec

  Status ReplicationControllerStatus
}

type ReplicationControllerSpec struct {
  ClusterSelector *LabelSelector

  Replicas *int32

  Selector *LabelSelector

  Template *PodTemplateSpec
}
```

We are almost certain that we are going to break this `ClusterSelector` semantics once we introduce spread/scheduling policies and such. So alternatively, what we should do is, introduce a `ClusterSpec` type and have the `ClusterSelector` information in that type. This `ClusterSpec` type can then embed all the cluster specific information and can be reused by all Ubernetes entities. It can look like this:

```
type ReplicationController struct {
  // Same as above
}

type ReplicationControllerSpec struct {
  Cluster *ClusterSpec

  Replicas *int32

  Selector *LabelSelector

  Template *PodTemplateSpec
}

type ClusterSpec struct {
  Name string
  Selector *LabelSelector
  // And any other cluster specific information we might want in the future.
}
```

`Name` can be optional. If the `Name` is specified we schedule all the replicas on to that cluster, otherwise we spread. Since we are going to creating sub-RCs in the spreading case, the sub-RCs will have the `Name` field populated.

This approach has some advantages:
1. It follows the semantics we have for pod scheduling with `NodeName` field in `PodSpec` and is consistent with it.
2. And it provides an area to persist federation level decisions.

But again, we don't have to do this in the first cut if it is too complicated. But I believe this won't be any more work than adding this information to annotations.

**Comment by mfanjie** [comment]
@madhusudancs thanks for your suggestion. We have offline sync and we agreed on the approach you suggested, @jianhuiz is making the structure change on kube-apiserver, I will resume my WIP later on.

**Comment by nelcy** [comment]
@madhusudancs @mfanjie 
Are we planning to reuse exact Kubernetes API Types for Ubernetes? The concern going forward will be overblowing these API types to include all Ubernetes specific & Kubernetes specific requirements. If not, shall we use different names or extensions to avoid confusion. 

Also, today in Kubernetes there is no binding or assignment/node selectors for replication controllers. They are at pod level.
My point overall it just adding ClusterSpec might not work.

Given that, I like the idea of just using annotations for now until we gather all requirements and converge on the final API data types.

**Comment by madhusudancs** [comment]
@neicy exact details still needs to be worked out. And may be others have thoughts too. Please feel free to pitch in. My suggestion was (and still is) to introduce new API types in Ubernetes API server. These types are actually new Go struct types in the code, but they mirror the fields in Kubernetes API server and add Ubernetes (cluster) specific information. @nikhiljindal also suggested that these types should have the same API group organization, i.e. group names and versions for equivalent types should be identical in Kubernetes and Ubernetes. Whether we want to embed the Kubernetes struct types or copy the field definitions is an open question.

> Also, today in Kubernetes there is no binding or assignment/node selectors for replication controllers. They are at pod level.

Sure. And how we want to schedule the RCs when these fields are specified is still an open design question.

> My point overall it just adding ClusterSpec might not work.

> Given that, I like the idea of just using annotations for now until we gather all requirements and converge on the final API data types.

I am almost always for taking a more conservative approach. But this is an exception. I am not entirely sure if we are promising Ubernetes API stability in v1.3. I would like to call this API alpha in the next release and give some time to let the API evolve. That's why I think we should take a more adventurous road now and get some feedback about it.

cc @quinton-hoole @lavalamp @nikhiljindal

**Comment by nelcy** [comment]
Thanks @madhusudancs for clarifying. I will check the PR and will ping if I have any questions.

**Comment by nikhiljindal** [comment]
To clarify, we will need the same group name, version and resource name for ubernetes and the corresponding kubernetes resource, if we want that users should be able to pass a kubernetes config file to kubectl talking to an ubernetes server (since the config file will have apiVersion and Kind fields).

We can have different Go types.
The go types can diverge over time if we copy them.

Re: Adding ClusterSpec (or ClusterSelector) in ReplicaSetSpec instead of in PodSpec: We will allow users to create a pod on ubernetes cluster, so we might need it in PodSpec anyway :)
ReplicaSet, Jobs and Deployments include PodSpec as well.
Also logically, ClusterSelector seems like a logical extension of NodeSelector to me, so it makes sense to keep them both at the same place in ubernetes go type.

Re: ClusterName: How does each cluster get a unique name? Does the admin specify it at the time of cluster registration? We can maybe use cluster UID?

I see that in the current proposal, the scheduling decision is stored in ReplicaSetSpec (ReplicaSetSpec.ClusterSpec.Name) in sub-RSs. I see that for pod, we store it in PodStatus (PodStatus.HostIP) - and not in PodSpec.

**Comment by nikhiljindal** [comment]
Also, for phase 1, I think we should just reuse the existing go type for replicasets and sub-replicasets in ubernetes and use annotations to add more info. We can use NodeSelector as proxy for ClusterSelector (We can add a label to all nodes indicating the cluster they are in).

We can do more once we have got it working end to end.

**Comment by mfanjie** [comment]
We (Ubernetes dev team) had some discussions and here are some outcomes, @jianhuiz is working on the go types definition. I summarized all relative topics in this issue and looking for further advises:
1. for backward compatibility consideration, federation RC will reuse current replicationcontroller go type, any additional info like clusterselector will be set as annotation, this is different from the design spec, so the RC spect will be changed from 
```
apiVersion: v1
kind: ReplicationController
metadata:
  name: nginx-controller
      clusterSelector: 
      name in (Foo, Bar)
```
to 
```
apiVersion: v1
kind: ReplicationController
metadata:
  name: nginx-controller
annotations:
  ubernetes.clusterselector: Foo, Bar (official annotation name can be discussed)
```

2. We need create a new type for subRC and it is basically a wrapper of RC, and currently we added a new ClusterSelectionSpec property as suggested above by @madhusudancs . subRC is created by ubernetes scheduler, and cluster controller watches it and dispatch the RC to corresponding cluster. It is fine for me for remove the new type and just use annotation to persist the target cluster info.  We do NOT need bind the clusterSelector to replicationcontroller so either way works. We need to make agreement here, I personally agree with the latter suggestion, just reuse the existing go type and use annotation to additional info.

3. What scheduler should do including:
a) watch RC on ube-apiserver
  - how to filter out scheduled ones or we do not need to?
  - current RC do not have status, so by default we just watch all RCs
b) check LabelSelector, if cluster:Foo is defined, then futher scheduling is not needed, create subRC(any other suggestion?)
c) LabelSelector is not defined, then 
  - check annotation of ubernetes.clusterselector
  - predicate, prioritize, select target and create subRC
d) behavior of creating subRC
  - clone from RC
  - add random suffix like pod
  - add annotation like tagetcluster to persist scheduling result
  - add labels to persist the parent federation RC name, so that cluster manager knows the relationships.

4. cluster is registered just like other k8s resources, to user need to define a unique name for all clusters. 

5. @nikhiljindal can you help me to understand more about your point on 
"Re: Adding ClusterSpec (or ClusterSelector) in ReplicaSetSpec instead of in PodSpec: We will allow users to create a pod on ubernetes cluster, so we might need it in PodSpec anyway :)"
My current understanding is we should stay on ReplicationController level, I am fine with storing the target clusterName in subRC annotation, but I think we should not care about ReplicaSet in federation level.

**Comment by nikhiljindal** [comment]
Re: ReplicationController and ReplicaSet - ReplicaSet is set to replace ReplicationController eventually. New resources (deployments, for example) work with ReplicaSet instead of ReplicationController. So I think, we should support ReplicaSet in ubernetes instead of ReplicationController.
LabelSelector is the only difference between ReplicaSet and ReplicationController.
All my comments talk about ReplicaSet but they can be applied to RC as well.


> We (Ubernetes dev team) had some discussions and here are some outcomes, @jianhuiz is working on the go types definition. I summarized all relative topics in this issue and looking for further advises:
> 
> ** Federation structure definition **
> 
> for backward compatibility consideration, federation RC will reuse current replicationcontroller go type, any additional info like clusterselector will be set as annotation, this is different from the design spec, so the RC spect will be changed from
> 
> apiVersion: v1
> kind: ReplicationController
> metadata:
>   name: nginx-controller
>       clusterSelector: 
>       name in (Foo, Bar)
> to
> 
> apiVersion: v1
> kind: ReplicationController
> metadata:
>   name: nginx-controller
> annotations:
>   federation.kubernetes.io/cluster-selector: Foo, Bar
> sub RC structure 
> We need create a new type for subRC and it is basically a wrapper of RC, and currently we added a new ClusterSelectionSpec property as suggested above by @madhusudancs . subRC is created by ubernetes scheduler, and cluster controller watches it and dispatch the RC to corresponding cluster. It is fine for me for remove the new type and just use annotation to persist the target cluster info. We do NOT need bind the clusterSelector to replicationcontroller so either way works. We need to make agreement here, I personally agree with the latter suggestion, just reuse the existing go type and use annotation to additional info.


Yes I agree with you. Reusing the existing go type seems good to me (for both RS and subRS).

> 
> What scheduler should do including
> 
> 1). watch RC on ube-apiserver**
> 
> how to filter out scheduled ones or we do not need to?

We can do that by detecting if appropriate subRSs exist for that RS.

> current RC do not have status, so by default we just watch all RCs

Yes we need to watch all RSs (for ex: to see if user has not deleted the RS or changed the number of replicas)

> 2). check LabelSelector, if cluster:Foo is defined, then futher scheduling is not needed, create subRC(any other suggestion?), discussion:
> 
> what labels users would define for labelselector: like only check labels with key "federation.kubernetes.io/cluster-name", and set the sub rc directly with target cluster name.

Sorry did not understand. ReplicaSets.Selector is for selecting the pods. It does not affect scheduling.

> 3). LabelSelector is not defined, then
> 
> check annotation of ubernetes.clusterselector
> predicate, prioritize, select target and create subRC
> 4). behavior of creating subRC
> 
> clone from RC
> add random suffix like pod
> add annotation like tagetcluster to persist scheduling result
> add labels to persist the parent federation RC name, so that cluster manager knows the relationships.

At other places we use `"kubernetes.io/created-by"` annotation.

> unique info: cluster name 
> cluster is registered just like other k8s resources, to user need to define a unique name for all clusters.
> 
> further question
> @nikhiljindal can you help me to understand more about your point on 
> "Re: Adding ClusterSpec (or ClusterSelector) in ReplicaSetSpec instead of in PodSpec: We will allow users to create a pod on ubernetes cluster, so we might need it in PodSpec anyway :)"
> My current understanding is we should stay on ReplicationController level, I am fine with storing the target clusterName in subRC annotation, but I think we should not care about ReplicaSet in federation level. 

Hopefully my comment above about ReplicaSet and ReplicationController made this clearer.

**Comment by madhusudancs** [comment]
I am fine with doing whatever is the easier/fastest to get this working end-to-end, i.e. adding annotations for now with Kubernetes Go types. But I am definitely not a fan of using nodeSelector as a proxy. It is fine if we want to prototype that way to prove everything works, but it shouldn't make it into a release, not even alpha. It just changes the semantic meaning of the field.

>> what labels users would define for labelselector: like only check labels with key "federation.kubernetes.io/cluster-name", and set the sub rc directly with target cluster name.

> Sorry did not understand. ReplicaSets.Selector is for selecting the pods. It does not affect scheduling.

Yes this. `labelSelector` shouldn't even be looked at for Ubernertes-level scheduling, not for phase 1 at least. Also, if `labelSelector` isn't supplied it is defaulted to pod labels in Kubernetes, so the RC should still be scheduled.

Also, we shouldn't encourage this form `federation.kubernetes.io/cluster-name` for labels/selectors. This form is for annotations which is mostly for tooling purposes. Labels and Selectors are for humans and it should be more readable.

> 3). LabelSelector is not defined, then

No, we should always look at `clusterSelector`, irrespective of what's in `labelSelector`.

> check annotation of ubernetes.clusterselector

I think looking at two different places will be a bit of additional work and we shouldn't be doing that for phase 1. For now, we should only look at `clusterSelector` and not the annotation. Let's use annotation only for recording the scheduler decisions for now. If necessary we can change that behavior later.


>> add labels to persist the parent federation RC name, so that cluster manager knows the relationships.

> At other places we use "kubernetes.io/created-by" annotation.

Yeah. Also, why is this in labels and the other information in annotations? I think both should be in annotations.

**Comment by chancez** [comment]
I hope drive-by comments aren't too frowned upon. I don't understand why the current proposal aims to modify the existing RC resource, instead of adding a new resource which templates RCs. (Ie: an ReplicationController is to Pods, as ClusterReplicationController is to ReplicationControllers). 

A very rough example of what I mean:
```yaml
apiVersion: v1alpha
kind: ClusterReplicationController
metadata:
  name: nginx-cluster-rc
  labels:
    app: nginx
spec:
  rc-selector:
    app: nginx
  template:
    apiVersion: v1
    kind: ReplicationController
    metadata:
      name: nginx-controller
    labels:
      app: nginx
    spec:
      containers:
        - image: nginx
          name: nginx
```

**Comment by madhusudancs** [comment]
@chancez That's what I proposed if you look at my comments earlier, specifically - https://github.com/kubernetes/kubernetes/issues/23269#issuecomment-200982500 and https://github.com/kubernetes/kubernetes/issues/23269#issuecomment-202656119.

But there is a lot of work to be done, esp. if we do not want to duplicate API server code and reuse as much as possible. And we do not want to block Ubernetes progress on that. @nikhiljindal can give more details.

We are definitely not proposing to modify existing RC types in Kubernetes APIs. But reusing them and adding Ubernetes specific information as annotations is acceptable for now I think. Adding annotations does not require modification to the type definitions. 

> Ie: an ReplicationController is to Pods, as ClusterReplicationController is to ReplicationControllers

Yes, and we don't even want the "Cluster" prefix there. That way users can submit the same config to either Kubernetes or Ubernetes API server without modifications (if it doesn't contain Ubernetes specific fields).

**Comment by madhusudancs** [comment]
@mfanjie: @nikhiljindal pointed out that making `clusterSelector` a field in RS would require us to create  these API types in Ubernetes. Since there is more work to be done before we can create those API types, let's add the cluster selector information to the annotations too. What you proposed for annotations is fine.

```yaml
apiVersion: v1
kind: ReplicationController
metadata:
  name: nginx-controller
annotations:
  federation.kubernetes.io/cluster-name: Foo, Bar
```

This is fine for now, except I changed `cluster-selector` to `cluster-name` for clarity.

**Comment by chancez** [comment]
@madhusudancs I see, sorry to add clutter. I was trying to follow along but couldn't for sure tell. That all sounds ideal, and you expressed my primary concerns. Makes sense. I'm really looking forward to this functionality getting in!

**Comment by madhusudancs** [comment]
@chancez no worries, more feedback is good. We want feedback :)

**Comment by mfanjie** [comment]
@madhusudancs about not using nodeSelector as proxy of ubernetes scheduling, I am with you, new Logic will be added as RC scheduling is a lot different from pod scheduling.

**Comment by mfanjie** [comment]
Had a sync up, and we will implement this based on ReplicaSet instead of controller manager.

**Comment by davidopp** [comment]
Explicitly naming clusters is totally reasonable for the first version. But longer-term, here is how I would suggest you select clusters:
(1) Have a Cluster resource in the Ubernetes control plane, one for each cluster (I think you are already planning to have it?)
(2) Put labels on the Cluster resources
(3) Have the annotation on ReplicaSet use NodeSelector
https://github.com/kubernetes/kubernetes/blob/master/pkg/api/v1/types.go#L1343
as the value, to select over the labels on the Cluster resources.

You will save yourself a LOT of work if you just reuse NodeSelector. (Note that this is the new NodeSelector type, not the old "nodeSelector" which is map[string]string.) We already have code that uses annotations on Pod to do this for pods, so it will require very little work.

**Comment by mfanjie** [comment]
@davidopp 
(1) @jianhuiz is responsible to cluster go type definition, and he already push the code in his wip, I think this part is not ready for review yet but you can take a look at (please note the latest comment from this issues were not addressed yet, like using RS vs. RC, and using annotation vs. new property)
https://github.com/XiaoningDing/UbernetesPOC/blob/ube-api-server-sub-rc-client/pkg/apis/controlplane/types.go#L70
(2) ObjectMeta is included for cluster type, so Labels are there.
(3) Thanks for the suggestion, if this is the right approach, I will check the code and see how I can reuse that.

**Comment by mfanjie** [comment]
@davidopp I looked further about the code piece referred by you, and yes, I totally agree such code can be reused, I just did not get the point before. I actually already created a ClusterSelector, ClusterSelectorRequirement, ClusterSelectorOperator, ClusterSelectorTerm similar with NodeSelector code. The code are almost same expect names, are you suggesting just leave the name as NodeSelector, and reuse them even we are not selecting cluster?

**Comment by davidopp** [comment]
That's a good question. I think the simplest thing to do is to clone NodeSelector, and rename it to ClusterSelector:

```
type ClusterSelector struct {
   NodeSelectorTerms []NodeSelectorTerm `json:"clusterSelectorTerms"`
}
```

(note that the only difference is NodeSelector became ClusterSelector, and nodeSelectorTerms became clusterSelectorTerms)

And then leave all the other types as-is. The Go types will be a little confusing (we will still have NodeSelectorTerm and NodeSelectorRequirement), but the Go types are not part of the API and are not user-visible, so this will not be confusing to users. The JSON field names do not have the word "node" in them (matchExpressions, key, operator, values), and that is the user-visible part.

Does that make sense?

BTW, just to be clear, for the first version I think it's fine to just select clusters by name, rather than selector. I was just suggestion a way to do it when you want something more sophisticated.

**Comment by mfanjie** [comment]
@davidopp exactly same thought here, I know now the cluster name is the only condition for federation scheduling, but still want make the scheduling more flexible so that we can easily add more logic in the future. I am thinking to covert the string array of clusterName to the format of clusterselector and then reuse created a predicate function call ReplicasSetMatchCluster to get the filtered cluster list. Let me know your further advises.

**Comment by davidopp** [comment]
I thnk @quinton-hoole should make this call. But personally I think it is simpler to just use cluster name for the first version. We can add selector approach later, and take the AND of the two if both are specified, so that it's not backward-incompatible. (Like we did when we added affinity on top of old nodeSelector that is map[string]string.)

**Comment by mfanjie** [comment]
Checking further of the code, as in P1 we don't support complex expression like node affinity terms, so it is hard to reuse existing NodeSelector code, to simplify, what I supposed to do is to iterated all cluster and check if the cluster name matches any of the annotation values.
I will send the wip for review anyway.

```
apiVersion: v1
kind: ReplicationController
metadata:
  name: nginx-controller
annotations:
  federation.kubernetes.io/cluster-name: Foo, Bar
```

**Comment by quinton-hoole** [comment]
fixes #21053

**Comment by quinton-hoole** [comment]
@mfanjie It sounds like all your questions have been addressed here?  I'm going to close this issue on that assumption, but feel free to re-open if you still need anything else from us.
